internal/planner/view.go

```diff
@@ -117,3 +107,142 @@ func (n *viewNode) Close() error {
 
 	return nil
 }
+
+func convertBetweenMaps(srcMap *core.DocumentMapping, dstMap *core.DocumentMapping, src core.Doc) core.Doc {
+	dst := dstMap.NewDoc()
+
+	srcRenderKeysByIndex := map[int]string{}
+	for _, renderKey := range srcMap.RenderKeys {
+		srcRenderKeysByIndex[renderKey.Index] = renderKey.Key
+	}
+
+	for underlyingName, srcIndexes := range srcMap.IndexesByName {
+		for _, srcIndex := range srcIndexes {
+			if srcIndex >= len(src.Fields) {
+				// Several system fields are not included in schema only types, and there is a mismatch somewhere
+				// that means we have to handle them here with a continue
+				continue
+			}
+
+			var dstName string
+			if key, ok := srcRenderKeysByIndex[srcIndex]; ok {
+				dstName = key
+			} else {
+				dstName = underlyingName
+			}
+
+			dstIndexes, dstHasField := dstMap.IndexesByName[dstName]
+			if !dstHasField {
+				continue
+			}
+
+			for _, dstIndex := range dstIndexes {
+				var srcValue any
+				if srcIndex < len(srcMap.ChildMappings) && srcMap.ChildMappings[srcIndex] != nil {
+					if dstIndex >= len(dstMap.ChildMappings) || dstMap.ChildMappings[dstIndex] == nil {
+						continue
+					}
+
+					switch inner := src.Fields[srcIndex].(type) {
+					case core.Doc:
+						srcValue = convertBetweenMaps(srcMap.ChildMappings[srcIndex], dstMap.ChildMappings[dstIndex], inner)
+
+					case []core.Doc:
+						dstInners := make([]core.Doc, len(inner))
+						for i, srcInnerDoc := range inner {
+							dstInners[i] = convertBetweenMaps(srcMap.ChildMappings[srcIndex], dstMap.ChildMappings[dstIndex], srcInnerDoc)
+						}
+						srcValue = dstInners
+					}
+				} else {
+					srcValue = src.Fields[srcIndex]
+				}
+
+				dst.Fields[dstIndex] = srcValue
+			}
+		}
+	}
+
+	return dst
+}
+
+// cachedViewFetcher is a planner node that fetches view items from a materialized cache.
+type cachedViewFetcher struct {
+	docMapper
+	documentIterator
+
+	def client.CollectionDefinition
+	p   *Planner
+
+	queryResults query.Results
+}
+
+var _ planNode = (*cachedViewFetcher)(nil)
+
+func (p *Planner) newCachedViewFetcher(
+	def client.CollectionDefinition,
+	mapping *core.DocumentMapping,
+) *cachedViewFetcher {
+	return &cachedViewFetcher{
+		def:       def,
+		p:         p,
+		docMapper: docMapper{mapping},
+	}
+}
+
+func (n *cachedViewFetcher) Init() error {
+	if n.queryResults != nil {
+		err := n.queryResults.Close()
+		if err != nil {
+			return err
+		}
+		n.queryResults = nil
+	}
+
+	prefix := core.NewViewCacheColPrefix(n.def.Description.RootID)
+
+	var err error
+	n.queryResults, err = n.p.txn.Datastore().Query(n.p.ctx, query.Query{
+		Prefix: prefix.ToString(),
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
```

Comment:
nitpick: `return err`

Reply:
For some reason I dislike that if it is not with the immediate result of another func call (e.g. `return foo()`), otherwise it looks like it is actually returning a non-nil error.  Leaving as is.